packages/next/src/server/pipe-readable.ts

```diff
@@ -1,107 +1,98 @@
+
 export function isAbortError(e: any): e is Error & { name: 'AbortError' } {
   return e?.name === 'AbortError'
 }
 
-/**
- * This is a minimal implementation of a Writable with just enough
- * functionality to handle stream cancellation.
- */
-export interface PipeTarget<R = any> {
-  /**
-   * Called when new data is read from readable source.
-   */
-  write: (chunk: R) => unknown
-
-  /**
-   * Always called once we read all data (if the writable isn't already
-   * destroyed by a client disconnect).
-   */
-  end: () => unknown
-
-  /**
-   * An optional method which is called after every write, to support
-   * immediately streaming in gzip responses.
-   */
-  flush?: () => unknown
-
-  /**
-   * The close event listener is necessary for us to detect an early client
-   * disconnect while we're attempting to read data. This must be done
-   * out-of-band so that we can cancel the readable (else we'd have to wait for
-   * the readable to produce more data before we could tell it to cancel).
-   */
-  on: (event: 'close', cb: () => void) => void
-
-  /**
-   * Allows us to cleanup our onClose listener.
-   */
-  off: (event: 'close', cb: () => void) => void
-
-  closed?: boolean
-}
+function createWriterFromResponse(
+  res: ServerResponse,
+  controller: AbortController
+): WritableStream<Uint8Array> {
+  let started = false
 
-export async function pipeReadable(
-  readable: ReadableStream<Uint8Array>,
-  writable: PipeTarget<Uint8Array>,
-  waitUntilForEnd?: Promise<void>
-) {
-  const reader = readable.getReader()
-  let readerDone = false
-  let writableClosed = false
-
-  // It's not enough just to check for `writable.destroyed`, because the client
-  // may disconnect while we're waiting for a read. We need to immediately
-  // cancel the readable, and that requires an out-of-band listener.
-  function onClose() {
-    writableClosed = true
-    writable.off('close', onClose)
-
-    // If the reader is not yet done, we need to cancel it so that the stream
-    // source's resources can be cleaned up. If a read is in-progress, this
-    // will also ensure the read promise rejects and frees our resources.
-    if (!readerDone) {
-      readerDone = true
-      reader.cancel().catch(() => {})
-    }
-  }
-  writable.on('close', onClose)
+  // Create a promise that will resolve once the response has drained. See
+  // https://nodejs.org/api/stream.html#stream_event_drain
+  let drained = Promise.withResolvers<void>()
+  res.on('drain', () => {
+    drained.resolve()
+  })
 
-  try {
-    while (true) {
-      const { done, value } = await reader.read()
-      readerDone = done
+  // Create a promise that will resolve once the response has finished. See
+  // https://nodejs.org/api/http.html#event-finish_1
+  const finished = Promise.withResolvers<void>()
+  res.once('finish', () => {
+    finished.resolve()
+
+    // If the finish event fires, it means we shouldn't block and wait for the
+    // drain event.
+    drained.resolve()
```

Comment:
The `finish` event is only fired if we close the client disconnects. We need to listen for the `close` event to unblock the drain.